src/main/java/UI/Parser.java

```diff
@@ -0,0 +1,142 @@
+package UI;
+
+import Storage.Storage;
+import Task.*;
+
+import java.time.format.DateTimeParseException;
+
+/**
+ * Parses user inputs and handles commands accordingly in a command-line interface.
+ */
+public class Parser {
+
+    public static Storage storage;
+
+    /**
+     * Parses the input command and delegates to the appropriate handler.
+     *
+     * @param input The full command input by the user.
+     * @param storage The storage instance to interact with file operations.
+     * @return true if the program should continue running, false if it should stop.
+     * @throws BotException If an unrecognized command is input.
+     */
+    public static boolean parse(String input, Storage storage) throws BotException {
+        Parser.storage = storage;
+        int spaceIndex = input.indexOf(" ");
+        String command = (spaceIndex != -1) ? input.substring(0, spaceIndex).toLowerCase() : input.toLowerCase();
+        String arguments = (spaceIndex != -1) ? input.substring(spaceIndex + 1).trim() : "";
+
+        switch (command) {
+        case "mark":
+        case "unmark":
+            handleMarkUnmark(command, arguments);
+            break;
+        case "todo":
+            handleToDo(arguments);
+            break;
+        case "deadline":
+            handleDeadline(arguments);
+            break;
+        case "event":
+            handleEvent(arguments);
+            break;
+        case "list":
+            handleList();
+            break;
+        case "delete":
+            handleDelete(arguments);
+            break;
+        case "bye":
+            handleBye();
+            return false;
+        case "find":
+            handleFind(arguments);
+            break;
+        default:
+            throw new BotException("I'm sorry, I don't recognize that command.");
+        }
+        return true;
+    }
+
+    private static void handleFind(String arguments) throws BotException{
+        if (arguments.isEmpty()) throw new BotException("Please provide a search term");
+        TaskList searchResultList = TaskList.mainTaskList.tasksContainingTerm(arguments);
+        searchResultList.printList();
+        if (searchResultList.getNumTasks() == 1) {
+            System.out.printf("There is 1 task containing the word \"%s\".\n", arguments);
+        } else {
+            System.out.printf("There are %s tasks containing the word \"%s\".\n", searchResultList.getNumTasks(), arguments);
+        }
+    }
+
+    private static void handleMarkUnmark(String command, String arguments) throws BotException {
+        if (arguments.isEmpty()) throw new BotException("Please provide a task number.");
+        int taskIndex = Integer.parseInt(arguments) - 1;
+        if (taskIndex < 0 || taskIndex >= TaskList.mainTaskList.getNumTasks()) throw new BotException("That task does not exist!");
```

Comment:
I think the if statement should be formatted as
```suggestion
if (taskIndex < 0 || taskIndex >= TaskList.mainTaskList.getNumTasks()) {
    throw new BotException("That task does not exist!");
}
```
so that the line isn't too long and it is easier to read.